### ISSUE TITLE
Random start up crash due to "Required key 'ust_router_address' is missing"

### ISSUE THREAD

**Issue by runephilosof** [phusion/passenger#2000]
## Issue report

https://superuser.com/questions/1261299/nginx-passenger-5-1-11-error-required-key-ust-router-address-is-missing

**Question 1: What is the problem?**
- What is the expected behavior?
`service apache2 reload` should work when the configs haven't been changed.

- What is the actual behavior?
Server crashed during log rotate
```
[Thu Nov 02 06:28:28.257749 2017] [passenger:error] [pid 1916:tid 140125707691904] *** Passenger could not be initialized because of this error: Required key 'ust_router_address' is missing
```
After restarting, it came back up showing no problems, no configs were changed. We don't even use Union Station.

- How can we reproduce it?
It only rarely happens, I have only seen it once. So I wouldn't know how to reproduce it.

```
[Thu Nov 02 06:28:26.915304 2017] [mpm_worker:notice] [pid 1916:tid 140125707691904] AH00297: SIGUSR1 received.  Doing graceful restart

[ N 2017-11-02 06:28:26.9165 5450/T5 age/Cor/CoreMain.cpp:588 ]: Signal received. Gracefully shutting down... (send signal 2 more time(s) to force shutdown)
[ N 2017-11-02 06:28:26.9166 5450/T1 age/Cor/CoreMain.cpp:1002 ]: Received command to shutdown gracefully. Waiting until all clients have disconnected...
[ N 2017-11-02 06:28:26.9166 5450/T1 age/Cor/CoreMain.cpp:917 ]: Checking whether to disconnect long-running connections for process 6186, application /data/sites/xxx
[ N 2017-11-02 06:28:26.9166 5450/T1 age/Cor/CoreMain.cpp:917 ]: Checking whether to disconnect long-running connections for process 1587, application /data/sites/xxx
[ N 2017-11-02 06:28:26.9167 5450/T1 age/Cor/CoreMain.cpp:917 ]: Checking whether to disconnect long-running connections for process 6233, application /data/sites/xxx
[ N 2017-11-02 06:28:26.9167 5450/T1 age/Cor/CoreMain.cpp:917 ]: Checking whether to disconnect long-running connections for process 13159, application /data/sites/xxx
[ N 2017-11-02 06:28:26.9167 5450/T1 age/Cor/CoreMain.cpp:917 ]: Checking whether to disconnect long-running connections for process 12379, application /data/sites/xxx
[ N 2017-11-02 06:28:26.9167 5450/T1 age/Cor/CoreMain.cpp:917 ]: Checking whether to disconnect long-running connections for process 5859, application /data/sites/xxx
[ N 2017-11-02 06:28:26.9167 5450/T1 age/Cor/CoreMain.cpp:917 ]: Checking whether to disconnect long-running connections for process 11229, application /data/sites/xxx
[ N 2017-11-02 06:28:26.9167 5450/T1 age/Cor/CoreMain.cpp:917 ]: Checking whether to disconnect long-running connections for process 5659, application /data/sites/xxx
[ N 2017-11-02 06:28:26.9167 5450/T1 age/Cor/CoreMain.cpp:917 ]: Checking whether to disconnect long-running connections for process 5912, application /data/sites/xxx
[ N 2017-11-02 06:28:26.9169 5450/Ta Ser/Server.h:886 ]: [ServerThr.2] Freed 128 spare client objects
[ N 2017-11-02 06:28:26.9170 5450/Ta Ser/Server.h:531 ]: [ServerThr.2] Shutdown finished
[ N 2017-11-02 06:28:26.9173 5450/T5 Ser/Server.h:886 ]: [ServerThr.1] Freed 128 spare client objects
[ N 2017-11-02 06:28:26.9174 5450/T5 Ser/Server.h:531 ]: [ServerThr.1] Shutdown finished

[ N 2017-11-02 06:28:26.9183 5456/T4 age/Ust/UstRouterMain.cpp:430 ]: Signal received. Gracefully shutting down... (send signal 2 more time(s) to force shutdown)
[ N 2017-11-02 06:28:26.9183 5450/Tc Ser/Server.h:886 ]: [ApiServer] Freed 0 spare client objects
[ N 2017-11-02 06:28:26.9184 5450/Tc Ser/Server.h:531 ]: [ApiServer] Shutdown finished
[ N 2017-11-02 06:28:26.9184 5456/T1 age/Ust/UstRouterMain.cpp:500 ]: Received command to shutdown gracefully. Waiting until all clients have disconnected...
[ N 2017-11-02 06:28:26.9184 5456/T6 Ser/Server.h:886 ]: [UstRouterApiServer] Freed 0 spare client objects
[ N 2017-11-02 06:28:26.9185 5456/T6 Ser/Server.h:531 ]: [UstRouterApiServer] Shutdown finished
[ N 2017-11-02 06:28:26.9186 5456/T4 Ser/Server.h:531 ]: [UstRouter] Shutdown finished
[ N 2017-11-02 06:28:26.9187 5456/T1 age/Ust/UstRouterMain.cpp:531 ]: Passenger UstRouter shutdown finished
[ N 2017-11-02 06:28:26.9195 5450/T1 age/Cor/CoreMain.cpp:917 ]: Checking whether to disconnect long-running connections for process 12379, application /data/sites/xxx
[ N 2017-11-02 06:28:27.1203 5450/T1 age/Cor/CoreMain.cpp:917 ]: Checking whether to disconnect long-running connections for process 6186, application /data/sites/xxx
[ N 2017-11-02 06:28:27.8223 5450/T1 age/Cor/CoreMain.cpp:917 ]: Checking whether to disconnect long-running connections for process 1587, application /data/sites/xxx
[ N 2017-11-02 06:28:28.0232 5450/T1 age/Cor/CoreMain.cpp:917 ]: Checking whether to disconnect long-running connections for process 6233, application /data/sites/xxx
[ N 2017-11-02 06:28:28.0232 5450/T1 age/Cor/CoreMain.cpp:917 ]: Checking whether to disconnect long-running connections for process 13159, application /data/sites/xxx
[Thu Nov 02 06:28:28.231740 2017] [ssl:warn] [pid 1916:tid 140125707691904] AH02292: Init: Name-based SSL virtual hosts only work for clients with TLS server name indication support (RFC 4366)
[ N 2017-11-02 06:28:28.2400 13411/T1 age/Wat/WatchdogMain.cpp:1267 ]: Starting Passenger watchdog...
[ N 2017-11-02 06:28:28.2495 13414/T1 age/Cor/CoreMain.cpp:1165 ]: Starting Passenger core...
[ N 2017-11-02 06:28:28.2498 13414/T1 age/Cor/CoreMain.cpp:249 ]: Passenger core running in multi-application mode.
[ N 2017-11-02 06:28:28.2559 13414/T1 age/Cor/CoreMain.cpp:905 ]: Passenger core online, PID 13414
[Thu Nov 02 06:28:28.257749 2017] [passenger:error] [pid 1916:tid 140125707691904] *** Passenger could not be initialized because of this error: Required key 'ust_router_address' is missing
[Thu Nov 02 06:28:28.257917 2017] [mpm_worker:notice] [pid 1916:tid 140125707691904] AH00292: Apache/2.4.7 (Ubuntu) OpenSSL/1.0.1f Phusion_Passenger/5.1.8 configured -- resuming normal operations
[Thu Nov 02 06:28:28.257931 2017] [core:notice] [pid 1916:tid 140125707691904] AH00094: Command line: '/usr/sbin/apache2'
[ W 2017-11-02 06:28:28.2588 13414/T1 age/Cor/CoreMain.cpp:1070 ]: Watchdog seems to be killed; forcing shutdown of all subprocesses

[ N 2017-11-02 06:28:28.2588 13414/T4 age/Cor/CoreMain.cpp:589 ]: Signal received. Gracefully shutting down... (send signal 2 more time(s) to force shutdown)
[ N 2017-11-02 06:28:28.9312 5450/T1 age/Cor/CoreMain.cpp:917 ]: Checking whether to disconnect long-running connections for process 5859, application /data/sites/xxx
[ N 2017-11-02 06:28:28.9313 5450/T1 age/Cor/CoreMain.cpp:917 ]: Checking whether to disconnect long-running connections for process 11229, application /data/sites/xxx
[ N 2017-11-02 06:28:29.3327 5450/T1 age/Cor/CoreMain.cpp:917 ]: Checking whether to disconnect long-running connections for process 5659, application /data/sites/xxx
[ N 2017-11-02 06:28:29.7346 5450/T1 age/Cor/CoreMain.cpp:917 ]: Checking whether to disconnect long-running connections for process 5912, application /data/sites/xxx
[ N 2017-11-02 06:28:30.1674 5450/T1 age/Cor/CoreMain.cpp:1068 ]: Passenger core shutdown finished
[Thu Nov 02 07:51:32.145109 2017] [mpm_worker:notice] [pid 1916:tid 140125707691904] AH00295: caught SIGTERM, shutting down
[Thu Nov 02 07:51:39.288277 2017] [ssl:warn] [pid 23987:tid 140253610289024] AH02292: Init: Name-based SSL virtual hosts only work for clients with TLS server name indication support (RFC 4366)
[ N 2017-11-02 07:51:39.2943 23989/T1 age/Wat/WatchdogMain.cpp:1267 ]: Starting Passenger watchdog...
[ N 2017-11-02 07:51:39.3004 23992/T1 age/Cor/CoreMain.cpp:1165 ]: Starting Passenger core...
[ N 2017-11-02 07:51:39.3006 23992/T1 age/Cor/CoreMain.cpp:249 ]: Passenger core running in multi-application mode.
[ N 2017-11-02 07:51:39.3050 23992/T1 age/Cor/CoreMain.cpp:905 ]: Passenger core online, PID 23992

[ N 2017-11-02 07:51:39.3060 23992/T7 age/Cor/CoreMain.cpp:589 ]: Signal received. Gracefully shutting down... (send signal 2 more time(s) to force shutdown)
[ N 2017-11-02 07:51:39.3061 23992/T1 age/Cor/CoreMain.cpp:1080 ]: Received command to shutdown gracefully. Waiting until all clients have disconnected...
[ N 2017-11-02 07:51:39.3062 23992/Ta Ser/Server.h:886 ]: [ServerThr.2] Freed 128 spare client objects
[ N 2017-11-02 07:51:39.3062 23992/Ta Ser/Server.h:531 ]: [ServerThr.2] Shutdown finished
[ N 2017-11-02 07:51:39.3063 23992/T7 Ser/Server.h:886 ]: [ServerThr.1] Freed 128 spare client objects
[ N 2017-11-02 07:51:39.3063 23992/T7 Ser/Server.h:531 ]: [ServerThr.1] Shutdown finished
[ N 2017-11-02 07:51:39.3066 23992/Tc Ser/Server.h:886 ]: [ApiServer] Freed 0 spare client objects
[ N 2017-11-02 07:51:39.3066 23992/Tc Ser/Server.h:531 ]: [ApiServer] Shutdown finished
[ N 2017-11-02 07:51:39.3400 23992/T1 age/Cor/CoreMain.cpp:1150 ]: Passenger core shutdown finished
[Thu Nov 02 07:51:39.966638 2017] [ssl:warn] [pid 24009:tid 140253610289024] AH02292: Init: Name-based SSL virtual hosts only work for clients with TLS server name indication support (RFC 4366)
[ N 2017-11-02 07:51:39.9727 24018/T1 age/Wat/WatchdogMain.cpp:1267 ]: Starting Passenger watchdog...
[ N 2017-11-02 07:51:39.9789 24021/T1 age/Cor/CoreMain.cpp:1165 ]: Starting Passenger core...
[ N 2017-11-02 07:51:39.9790 24021/T1 age/Cor/CoreMain.cpp:249 ]: Passenger core running in multi-application mode.
[ N 2017-11-02 07:51:39.9834 24021/T1 age/Cor/CoreMain.cpp:905 ]: Passenger core online, PID 24021
[Thu Nov 02 07:51:39.984771 2017] [mpm_worker:notice] [pid 24009:tid 140253610289024] AH00292: Apache/2.4.7 (Ubuntu) OpenSSL/1.0.1f Phusion_Passenger/5.1.11 configured -- resuming normal operations
[Thu Nov 02 07:51:39.984816 2017] [core:notice] [pid 24009:tid 140253610289024] AH00094: Command line: '/usr/sbin/apache2'
```

**Question 2: Passenger version and integration mode**:

Installed through `deb https://oss-binaries.phusionpassenger.com/apt/passenger trusty main`

**Question 3: OS or Linux distro, platform (including version)**:

Ubuntu 14.04.5 LTS
```
libapache2-mod-passenger       1:5.1.11-1~trusty1   amd64
passenger                      1:5.1.11-1~trusty1   amd64
passenger-dev                  1:5.1.11-1~trusty1   amd64
passenger-doc                  1:5.1.11-1~trusty1   all
```

**Question 4: Passenger installation method**:

Your answer:
[ ] RubyGems + Gemfile
[ ] RubyGems, no Gemfile
[x] Phusion APT repo
[ ] Phusion YUM repo
[ ] OS X Homebrew
[ ] source tarball
[ ] Other, please specify:

**Question 5: Your app's programming language (including any version managers) and framework (including versions)**:

Ruby (2.4.0, 2.1.7, and 2.4.1) and Rails versions are different for each app by use of rbenv.
```
<IfModule mod_passenger.c>
  PassengerRoot "/usr/lib/ruby/vendor_ruby/phusion_passenger/locations.ini"
  PassengerDefaultRuby "/opt/rbenv/shims/ruby"
```

**Question 6: Are you using a PaaS and/or containerization? If so which one?**
- For example: Heroku, Amazon Container Services, Docker 1.9 with an image based on passenger-docker

Your answer:
No

**Question 7: Anything else about your setup that we should know?**

Your answer:

**Comment by OnixGH** [comment]
The issue you linked to was determined to be due to a partial upgrade of Passenger (solved by updating the `nginx-extras` package).

I may be wrong, but my first guess would be that it's also related to a Passenger upgrade on your system. I think `service apache2 reload` doesn't reload modules, so if you upgrade Passenger and only reload Apache, then that would create a situation where the old Passenger Apache module is talking to the upgraded Passenger core. This would also explain your observation that the problem went away after a full Apache restart.

Normally this old module / new core mismatch wouldn't manifest as an issue because we don't often change something that affects the interaction between those two.

Has the problem been seen on servers that were restarted since the upgrade to 5.1.11?

**Comment by OnixGH** [comment]
Update: I took some time to test this, and indeed it's reproduceable.

If you upgrade Passenger and you just do a `service apache2 reload`, the module won't be reloaded and you'll get the old module talking to the new Passenger core. The logs also reflect this, for example when I upgraded from 5.1.5 to 5.1.11 and did a reload, the logs still show the old module being loaded:
```
[Thu Nov 02 14:38:18.194463 2017] [mpm_event:notice] [pid 26906:tid 140672939677568] AH00489: Apache/2.4.18 (Ubuntu) Phusion_Passenger/5.1.5 configured -- resuming normal operations
```

We'll have to make sure the docs (and maybe release notes) cover this.

**Comment by johnl** [comment]
We've seen this upgrading packages from passenger 5.0 to 5.1 on Ubuntu trusty.

The upgrade completes and the package post install scripts *reload* apache:

`Setting up passenger (1:5.1.11-1~trusty1) ...
Setting up passenger-dev (1:5.1.11-1~trusty1) ...
Setting up libapache2-mod-passenger (1:5.1.11-1~trusty1) ...
 * Reloading web server apache2                                                                                                                                                                                                                                 * 
`

The parent apache2 process keep running but it's child apache2 processes are restarted, and then passenger fails to spawn:

`[Tue Nov 07 13:09:51.264246 2017] [mpm_worker:notice] [pid 20757:tid 3073951552] AH00297: SIGUSR1 received.  Doing graceful restart

[ 2017-11-07 13:09:51.2664 20764/b5cffb40 age/Cor/CoreMain.cpp:532 ]: Signal received. Gracefully shutting down... (send signal 2 more time(s) to force shutdown)
[ 2017-11-07 13:09:51.2665 20764/b6aa9740 age/Cor/CoreMain.cpp:901 ]: Received command to shutdown gracefully. Waiting until all clients have disconnected...
[ 2017-11-07 13:09:51.2665 20764/b6aa9740 age/Cor/CoreMain.cpp:819 ]: Checking whether to disconnect long-running connections for process 21176, application /home/controlpanel/controlpanel/current (production)
[ 2017-11-07 13:09:51.2666 20764/b6aa9740 age/Cor/CoreMain.cpp:819 ]: Checking whether to disconnect long-running connections for process 21181, application /home/controlpanel/controlpanel/current (production)
[ 2017-11-07 13:09:51.2668 20764/b5cffb40 Ser/Server.h:817 ]: [ServerThr.1] Freed 128 spare client objects
[ 2017-11-07 13:09:51.2669 20764/b5cffb40 Ser/Server.h:464 ]: [ServerThr.1] Shutdown finished
[ 2017-11-07 13:09:51.2679 20764/b4c7cb40 Ser/Server.h:817 ]: [ApiServer] Freed 0 spare client objects
[ 2017-11-07 13:09:51.2679 20764/b4c7cb40 Ser/Server.h:464 ]: [ApiServer] Shutdown finished

[ 2017-11-07 13:09:51.2680 20769/b67ffb40 age/Ust/UstRouterMain.cpp:422 ]: Signal received. Gracefully shutting down... (send signal 2 more time(s) to force shutdown)
[ 2017-11-07 13:09:51.2680 20769/b6a80740 age/Ust/UstRouterMain.cpp:492 ]: Received command to shutdown gracefully. Waiting until all clients have disconnected...
[ 2017-11-07 13:09:51.2681 20764/b54feb40 Ser/Server.h:817 ]: [ServerThr.2] Freed 128 spare client objects
[ 2017-11-07 13:09:51.2681 20769/b67ffb40 Ser/Server.h:464 ]: [UstRouter] Shutdown finished
[ 2017-11-07 13:09:51.2681 20764/b54feb40 Ser/Server.h:464 ]: [ServerThr.2] Shutdown finished
[ 2017-11-07 13:09:51.2682 20769/b5bffb40 Ser/Server.h:817 ]: [UstRouterApiServer] Freed 0 spare client objects
[ 2017-11-07 13:09:51.2682 20769/b5bffb40 Ser/Server.h:464 ]: [UstRouterApiServer] Shutdown finished
[ 2017-11-07 13:09:51.2686 20769/b6a80740 age/Ust/UstRouterMain.cpp:523 ]: Passenger UstRouter shutdown finished
[ 2017-11-07 13:09:51.2712 20764/b6aa9740 age/Cor/CoreMain.cpp:819 ]: Checking whether to disconnect long-running connections for process 21176, application /home/controlpanel/controlpanel/current (production)
[ 2017-11-07 13:09:51.2713 20764/b6aa9740 age/Cor/CoreMain.cpp:819 ]: Checking whether to disconnect long-running connections for process 21181, application /home/controlpanel/controlpanel/current (production)
[ N 2017-11-07 13:09:51.3217 29400/T1 age/Wat/WatchdogMain.cpp:1267 ]: Starting Passenger watchdog...
[ N 2017-11-07 13:09:51.3359 29403/T1 age/Cor/CoreMain.cpp:1165 ]: Starting Passenger core...
[ N 2017-11-07 13:09:51.3368 29403/T1 age/Cor/CoreMain.cpp:249 ]: Passenger core running in multi-application mode.
[ N 2017-11-07 13:09:51.3466 29403/T1 age/Cor/CoreMain.cpp:905 ]: Passenger core online, PID 29403
[Tue Nov 07 13:09:51.361117 2017] [passenger:error] [pid 20757:tid 3073951552] *** Passenger could not be initialized because of this error: Required key 'ust_router_address' is missing
[Tue Nov 07 13:09:51.363491 2017] [mpm_worker:notice] [pid 20757:tid 3073951552] AH00292: Apache/2.4.7 (Ubuntu) Phusion_Passenger/5.0.30 OpenSSL/1.0.1f configured -- resuming normal operations
[Tue Nov 07 13:09:51.363534 2017] [core:notice] [pid 20757:tid 3073951552] AH00094: Command line: '/usr/sbin/apache2'
[ W 2017-11-07 13:09:51.3646 29403/T1 age/Cor/CoreMain.cpp:1070 ]: Watchdog seems to be killed; forcing shutdown of all subprocesses

[ N 2017-11-07 13:09:51.3647 29403/T5 age/Cor/CoreMain.cpp:589 ]: Signal received. Gracefully shutting down... (send signal 2 more time(s) to force shutdown)
`
this leaves apache2 running but with no passenger - requests to the app get a 404.

A full *restart* of the apache2 service (rather than a reload) fixes this.

We've hit this after an automatic upgrade too, which broke things unexpectedly!

We've been doing this for quite some time and not experienced such a problem (though it's possible the version upgrade wasn't taking effect but wasn't causing more serious problems until now).

So the package post install scripts should probably be doing a restart rather than a reload, at least if it's an upgrade that will cause such a problem (so you could check the version step in the script and restart conditionally perhaps)

**Comment by OnixGH** [comment]
@johnl good to know that your symptoms also fit the hypothesis for this issue.

And you're right on the full restart, we were looking into why that isn't done already.

**Comment by OnixGH** [comment]
Turns out it was fixed right after 5.1.11 was released (https://github.com/phusion/passenger_apt_automation/commit/aa3cad2ceb9a041ad0c0f4ef1b79c3ae4c00dbf0), so this issue should be solved for future releases.
